src/thumbnailer_compare.cc

```diff
@@ -0,0 +1,125 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "thumbnailer.h"
+
+namespace libwebp {
+
+Thumbnailer::Status Thumbnailer::AnimData2Pictures(
+    WebPData* const webp_data,
+    std::vector<std::unique_ptr<WebPPicture, void (*)(WebPPicture*)>>* const
+        pics) {
+  std::unique_ptr<WebPAnimDecoder, void (*)(WebPAnimDecoder*)> dec(
+      WebPAnimDecoderNew(webp_data, NULL), WebPAnimDecoderDelete);
+  if (dec == NULL) {
+    std::cerr << "Error parsing image." << std::endl;
+    return kMemoryError;
+  }
+
+  WebPAnimInfo anim_info;
+  if (!WebPAnimDecoderGetInfo(dec.get(), &anim_info)) {
+    std::cerr << "Error getting global info about the animation";
+    return kMemoryError;
+  }
+
+  const int width = anim_info.canvas_width;
+  const int height = anim_info.canvas_height;
+
+  while (WebPAnimDecoderHasMoreFrames(dec.get())) {
+    uint8_t* frame_rgba;
+    int timestamp;
+    if (!WebPAnimDecoderGetNext(dec.get(), &frame_rgba, &timestamp)) {
+      std::cerr << "Error decoding frame." << std::endl;
+      return kMemoryError;
+    }
+    pics->emplace_back(new WebPPicture, WebPPictureFree);
+    WebPPicture* pic = pics->back().get();
+    if (!WebPPictureInit(pic)) return kMemoryError;
+    pic->use_argb = 1;
+    pic->width = width;
+    pic->height = height;
+    if (!WebPPictureImportRGBA(pic, frame_rgba, width * 4)) {
+      return kMemoryError;
+    }
+  }
+  return kOk;
+}
+
+Thumbnailer::Status Thumbnailer::AnimData2PSNR(WebPData* const webp_data,
+                                               std::vector<float>* const psnr) {
+  std::vector<std::unique_ptr<WebPPicture, void (*)(WebPPicture*)>> pics;
+  if (AnimData2Pictures(webp_data, &pics) != kOk) return kMemoryError;
+  if (pics.size() != frames_.size()) return kMemoryError;
+
+  for (int i = 0; i < pics.size(); ++i) {
+    auto& frame = frames_[i];
+    auto& pic = pics[i];
+
+    float distortion_result[5];
+
+    if (!WebPPictureDistortion(&frame.pic, pic.get(), 0, distortion_result)) {
+      break;
+    } else {
+      psnr->push_back(distortion_result[4]);  // PSNR-all.
+    }
+  }
+
+  if (psnr->size() != pics.size()) return kMemoryError;
+  return kOk;
+}
+
+Thumbnailer::Status Thumbnailer::CompareThumbnail(WebPData* const webp_data_ref,
+                                                  WebPData* const webp_data) {
+  std::vector<float> psnr_ref, psnr, psnr_diff;
+  if (AnimData2PSNR(webp_data_ref, &psnr_ref) != kOk) return kMemoryError;
+  if (AnimData2PSNR(webp_data, &psnr) != kOk) return kMemoryError;
+  if (psnr_ref.size() != psnr.size()) return kMemoryError;
+
+  // WebPData doesn't contain any frames.
+  if (psnr.empty()) return kOk;
+
+  const int frame_count = psnr.size();
+
+  for (int i = 0; i < frame_count; ++i) {
+    psnr_diff.push_back(psnr[i] - psnr_ref[i]);
+  }
+
+  std::sort(psnr_diff.begin(), psnr_diff.end());
```

Comment:
Sorting is unnecessary, `std::minmax_element()` is enough and faster

Reply:
I sorted `psnr_diff` to get the median.

Reply:
Oh right I forgot. You have `std::nth_element()` for that but do not bother indeed, sorting is fine.